### ISSUE TITLE
Requesting native support for android and arm (raspberrypi)

### ISSUE THREAD

**Issue by s4kibs4mi** [xerial/sqlite-jdbc#172]
It would be excellent if you could add native support for android and arm. I have developed an [ORM library for kotlin](https://github.com/s4kibs4mi/PultusORM) using sqlite-jdbc but to use the library in android and arm it needs native support for these platforms.

**Comment by xerial** [comment]
arm native library is already supported in sqlite-jdbc-3.14.2.1. If it doesn't work for you, could you post a stack trace message when using this version? This message will have os.name and os.arch information of your JVM.

**Comment by s4kibs4mi** [comment]
@xerial I have added the latest version and its working like a charm. I have tested on android.Great job man. :D

**Comment by s4kibs4mi** [comment]
When I am running it on android its showing
`java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/ninja.sakib.androidmultiprocesstest-1/base.apk"],nativeLibraryDirectories=[/vendor/lib, /system/lib]]] couldn't find "libsqlitejdbc.so"`

I have added `compile group: 'org.xerial', name: 'sqlite-jdbc', version: '3.14.2.1'`

**Comment by xerial** [comment]
weird. It should show: `Failed to load native library:  .... `
https://github.com/xerial/sqlite-jdbc/blob/c591bfde13a208be7838a628e924b65b651e045c/src/main/java/org/sqlite/SQLiteJDBCLoader.java#L230

Could you show your System property values of `os.name` and `os.arch`?

**Comment by s4kibs4mi** [comment]
Here my devices system properties as requested,
`Name: Linux`
`Arch: armv7l`

**Comment by xerial** [comment]
I'm looking for the  cross compiler for building this:
https://github.com/dockcross/dockcross 
I'll try building a native library for armv7l with dockcross/linux-armv7 image.

**Comment by s4kibs4mi** [comment]
That would be great.

**Comment by g1ko** [comment]
Please  add  native library for armv5 (armv5tel) as well.

**Comment by xerial** [comment]
Hi guys, 

I would like to ask a question for arm support. What values are shown if you run the following Java  code?
```
System.getProperty("os.name")
System.getProperty("os.arch")
```

We need these information to add support armv7 and arvm5, etc.

**Comment by g1ko** [comment]
os.name = Linux
os.arch   = arm

**Comment by xerial** [comment]
@g1ko 
Thanks. That value `arm` means we need a reliable way to distinguish various arm versions (e.g., armv5, armv7l, etc.)

**Comment by xerial** [comment]
This might be related for detecting arm CPU types: http://stackoverflow.com/questions/14859954/android-os-arch-output-for-arm-mips-x86

**Comment by g1ko** [comment]
Also It would be great to  have the library for powerpc.
os.name = Linux
os.arch = ppc

#java -version
java version "1.8.0_06"
Java(TM) SE Embedded Runtime Environment (build 1.8.0_06-b23, headless)
Java HotSpot(TM) Embedded Client VM (build 25.6-b23, mixed mode)

#uname -a
Linux MybookLive 4.1.17-mbl #1 Sat Feb 13 10:46:06 UTC 2016 ppc GNU/Linux

#cat /proc/cpuinfo
processor	: 0
cpu		: APM821XX
clock		: 800.000008MHz
revision	: 28.130 (pvr 12c4 1c82)
bogomips	: 1600.00
timebase	: 800000008
platform	: PowerPC 44x Platform
model		: amcc,apollo3g
Memory		: 256 MB

**Comment by xerial** [comment]
ppc support would be relatively easy.

**Comment by g1ko** [comment]
I did try to build the .jar by myself but no dice. 
I used Ubuntu 14.04LTS VM as my cross  development tool with docker-engine installed, git-cloned latest (3.15.1) version.  I run "make linux-arm". The build process completed with no errors, the .so file produced in target/sqlite-...Linux-arm folder but  .jar nowhere . 
Is it ok/expected?

**Comment by g1ko** [comment]
Answering to myself: issuing just "make" did the trick, .jar  BUILD SUCCESSFUL.

**Comment by g1ko** [comment]
>ppc support would be relatively easy.
Is that real before Xmas?  

Looking forward.

**Comment by alecmuffett** [comment]
Hi Taro; I'm trying to compile on Raspbian / Raspberry Pi 3 model B -- on the basis of old documentation which says that the ARM is not supported. You mention above that ARM is supported (which is nice :-) ) but builds are still failing with lack of "dockcross". 

I'll admit I'm not really sure what is going on, but I am guessing that it's no longer possible to compile natively on anything other than a x86 machine, using cross-compilation for ARM.  Would that be correct? This is compared to `README_BUILD.md` and the advice that `f you want to use the native library for your OS, build the source from scratch.` ?

```
$ env JAVA_HOME=/usr/lib/jvm/jdk-8-oracle-arm32-vfp-hflt make
./docker/dockcross-windows-x86 bash -c 'make clean-native native CROSS_PREFIX=i686-w64-mingw32.static- OS_NAME=Windows OS_ARCH=x86'
./docker/dockcross-windows-x86: line 182: docker: command not found
Makefile:102: recipe for target 'win32' failed
```

**Comment by xerial** [comment]
@g1ko I'll find a time for this :)

@alecmuffett 
I have removed the cross compiler based build script from Makefile. I can add some make target so that you can build native libraries by yourself.

The current status is: we are delegating Travis CI to build all of the supported native libraries using dockcross (or cross compiler images I have prepared for Linux). TravisCI now supports docker, so basically we don't need to build native libraries by ourselves if your platform is supported in the master branch. 

Currently TravisCI pushes a snapshot version that contains docker-built native libraries every time if there is a new commit to the master branch:
https://oss.sonatype.org/content/repositories/snapshots/org/xerial/sqlite-jdbc/3.15.2-SNAPSHOT/

This should reduce the burden of preparing cross compilers (You don't need to have docker nor linux box)

We also should fix the README.md to include ARM support :)

**Comment by psini** [comment]
Are any news for Android  native lib build using dockcross ? 
Thank you in advance.

**Comment by xerial** [comment]
#193 addresses this (android, raspberry (armv6) support)

**Comment by xerial** [comment]
@s4kibs4mi @psini @g1ko 
Here is the latest binary that includes android, armv6, armv7 support:
https://oss.sonatype.org/content/repositories/snapshots/org/xerial/sqlite-jdbc/3.16.1-SNAPSHOT/sqlite-jdbc-3.16.1-20170106.232609-17.jar

Let me know if you find any problem with this version. Thanks.

**Comment by xerial** [comment]
Released sqlite-jdbc 3.16.1, which probably fixes this problem. If you find any problem, create another ticket and I will work on the fix when releaseing 3.16.2.
